### PR TITLE
fix: EPERM error on windows installation

### PR DESCRIPTION
**Pull request by MakakWasTaken**
fix: EPERM error on windows installation

Moved the pull request from here: https://github.com/biomejs/biome/pull/735

### CODE REVIEW

**Comment by nhedger** [comment]
Thanks for reopening here @MakakWasTaken 🙏

File: src/main.ts [hunk 1]
```diff
@@ -141,11 +136,74 @@ export async function activate(context: ExtensionContext) {
...
+
+		let tmpDestination: string;
+
+		try {
+			tmpDestination = path.resolve(
+				os.tmpdir(),
+				`./biome${process.platform === "win32" ? ".exe" : ""}`,
+			);
+			outputChannel.appendLine(`Copying file to tmp folder: ${tmpDestination}`);
+			copyFileSync(server.command, tmpDestination);
```
Comment (nhedger) [inline src/main.ts:151]: I think we should prefer using the `workspace.fs` api to copy the file, if possible.

https://code.visualstudio.com/api/references/vscode-api#FileSystem

**Comment by nhedger** [comment]
Wondering if we should use `storageUri` instead to store the copy of the biome binary

> The uri of a workspace specific directory in which the extension can store private state. The directory might not exist and creation is up to the extension. However, the parent directory is guaranteed to be existent. The value is `undefined` when no workspace nor folder has been opened.

https://code.visualstudio.com/api/references/vscode-api#ExtensionContext

**Comment by MakakWasTaken** [comment]
> Wondering if we should use `storageUri` instead of the temp dir to store the copy of the biome binary
> 
> > The uri of a workspace specific directory in which the extension can store private state. The directory might not exist and creation is up to the extension. However, the parent directory is guaranteed to be existent. The value is `undefined` when no workspace nor folder has been opened.
> 
> https://code.visualstudio.com/api/references/vscode-api#ExtensionContext

Might be smart, would allow for a workspace-based tmp file.

File: src/main.ts [hunk 2]
```diff
@@ -26,7 +26,10 @@ import { StatusBar } from "./statusBar";
 import { setContextValue } from "./utils";
 
 import resolveImpl = require("resolve/async");
+import { copyFileSync } from "fs";
 import { createRequire } from "module";
+import * as os from "os";
+import * as path from "path";
```
Comment (nhedger) [inline src/main.ts:32]:
```suggestion

```

File: src/main.ts [hunk 3]
```diff
@@ -26,7 +26,10 @@ import { StatusBar } from "./statusBar";
 import { setContextValue } from "./utils";
 
 import resolveImpl = require("resolve/async");
+import { copyFileSync } from "fs";
```
Comment (nhedger) [inline src/main.ts:29]:
```suggestion

```

File: src/main.ts [hunk 4]
```diff
@@ -141,11 +136,73 @@ export async function activate(context: ExtensionContext) {
...
+	await reloadClient();
+
+	// Best way to determine package updates. Will work for npm, yarn, pnpm and bun. (Might work for more files also).
+	// It is not possible to listen node_modules, because it is usually gitignored.
+	const watcher = workspace.createFileSystemWatcher("**/*lock*");
+	context.subscriptions.push(
+		watcher.onDidChange(async () => {
+			try {
+				// When the lockfile changes, reload the biome executable.
+				outputChannel.appendLine("Reloading biome executable..");
```
Comment (nhedger) [inline src/main.ts:192]:
```suggestion
				outputChannel.appendLine("Reloading biome executable.");
```

**Comment by MakakWasTaken** [comment]
Is there more missing on this PR?

**Comment by nhedger** [comment]
Hey, I just want to make sure it works as expected on all platforms.

I'll test it this week and hopefully merge it before this weekend.

**Comment by MakakWasTaken** [comment]
Alright, just wanted to make sure I didn't miss anything.

**Comment by nhedger** [comment]
Currently, the code will copy the binary no matter where it comes from, but we only want to copy it when it comes from the project's dev dependencies.

Particularly, we do **not** want to copy the binary when:
- the user is using the bundled biome binary
- the binary has been configured with the `lspBin` setting
- the binary path has been set with the `DEBUG_SERVER_PATH` env variable

**Comment by MakakWasTaken** [comment]
I ended up adding a new property called `workspaceDependency` to the `getServerPath` function. An alternative way of doing this could be looking after `node_modules/@biomejs` or something like that in the path. There is however a possibility of users setting either the `lspBin` or `DEBUG_SERVER_PATH` to a path including this, which would mess up the result of the function.

**Comment by nhedger** [comment]
I was able to confirm it works on macOS.

I'll be checking the other systems shortly.

I've slightly reworded some logs and added a missing await statement to prevent the extension from starting the server before the file is copied over.

**Comment by veaba** [comment]
+1

Can not find out Biome in formatter list

**Comment by nhedger** [comment]
> +1
> 
> Can not find out Biome in formatter list

Not sure what you mean here.

**Comment by veaba** [comment]
> > Can not find out Biome in formatter list

Biome is now found in the list of formatters.

maybe I should restart vscode, 

image

**Comment by nhedger** [comment]
This is a pull request to address another issue.

Please either open a new issue or hop on Discord and we'll help you figure it out.

**Pull request merged**